src/reactors.js

```diff
@@ -1,27 +1,21 @@
 import _ from 'lodash/fp'
-import { lookup } from './util/tree'
 
 // TODO check type, etc
 let hasContext = node => node.context
+let hadValue = previous => {
+  if (previous && previous.hasValue === undefined)
+    throw Error('Node was never validated')
+  return previous && previous.hasValue && !previous.error
```

Comment:
Hi! So, questions:
- Can has value be falsy? If it can: then we need to change this line, if it can't: then we can simplify line 6.
- We might reduce this a bit with _.get

Reply:
`hasValue` should generally only be `true` or `false` by the time it reaches this line - but if it's undefined it means that validation never ran. While that shouldn't happen, this PR is a fundamental shift to assume that validation runs one time at the start of the dispatch. While authoring this PR, that line helped me diagnose places where that assumption was invalid - and leaving it in should help us catch issues like that in the future.

Reply:
Interestingly, `get` introduces a bug, because `_.get('hasValue', previous)` when previous itself is null would return undefined and trigger the exception - but if previous itself is null, the correct behavior is to return false. I think it could be made a bit clearer though 🤔